1.6.0 error: undefined method `library_of_name' for nil:NilClass (NoMethodError)

This just started today, so I suspect it is an issue with master@439d668.

```
Using built-in specs.
COLLECT_GCC=g++
COLLECT_LTO_WRAPPER=/usr/lib/gcc/x86_64-linux-gnu/11/lto-wrapper
OFFLOAD_TARGET_NAMES=nvptx-none:amdgcn-amdhsa
OFFLOAD_TARGET_DEFAULT=1
Target: x86_64-linux-gnu
Configured with: ../src/configure -v --with-pkgversion='Ubuntu 11.3.0-1ubuntu1~22.04.1' --with-bugurl=file:///usr/share/doc/gcc-11/README.Bugs --enable-languages=c,ada,c++,go,brig,d,fortran,objc,obj-c++,m2 --prefix=/usr --with-gcc-major-version-only --program-suffix=-11 --program-prefix=x86_64-linux-gnu- --enable-shared --enable-linker-build-id --libexecdir=/usr/lib --without-included-gettext --enable-threads=posix --libdir=/usr/lib --enable-nls --enable-bootstrap --enable-clocale=gnu --enable-libstdcxx-debug --enable-libstdcxx-time=yes --with-default-libstdcxx-abi=new --enable-gnu-unique-object --disable-vtable-verify --enable-plugin --enable-default-pie --with-system-zlib --enable-libphobos-checking=release --with-target-system-zlib=auto --enable-objc-gc=auto --enable-multiarch --disable-werror --enable-cet --with-arch-32=i686 --with-abi=m64 --with-multilib-list=m32,m64,mx32 --enable-multilib --with-tune=generic --enable-offload-targets=nvptx-none=/build/gcc-11-aYxV0E/gcc-11-11.3.0/debian/tmp-nvptx/usr,amdgcn-amdhsa=/build/gcc-11-aYxV0E/gcc-11-11.3.0/debian/tmp-gcn/usr --without-cuda-driver --enable-checking=release --build=x86_64-linux-gnu --host=x86_64-linux-gnu --target=x86_64-linux-gnu --with-build-config=bootstrap-lto-lean --enable-link-serialization=2
Thread model: posix
Supported LTO compression algorithms: zlib zstd
gcc version 11.3.0 (Ubuntu 11.3.0-1ubuntu1~22.04.1) 
Fetching https://github.com/Arduino-CI/arduino_ci
Resolving dependencies...
Fetching gem metadata from https://rubygems.org/....
Fetching os 1.1.4
Fetching rubyzip 1.3.0
Installing os 1.1.4
Installing rubyzip 1.3.0
Using bundler 2.4.10
Using arduino_ci 1.6.0 from https://github.com/Arduino-CI/arduino_ci (at master@439d668)
Bundle complete! 1 Gemfile dependency, 4 gems now installed.
Bundled gems are installed into `./vendor/bundle`
Attempting to download Arduino 0.29.0 package with open-uri....
Extracting archive with tar
Cloning into 'Adafruit_BusIO'...
Cloning into 'Adafruit_MAX31865_library'...
Cloning into 'Ethernet'...
Cloning into 'LiquidCrystal'...
Cloning into 'RTClib'...
Cloning into 'Keypad'...
Cloning into 'PID'...
Cloning into 'SdFat'...
Using os 1.1.4
Using rubyzip 1.3.0
Using arduino_ci 1.6.0 from https://github.com/Arduino-CI/arduino_ci (at master@439d668)
Using bundler 2.4.10
Bundle complete! 1 Gemfile dependency, 4 gems now installed.
Bundled gems are installed into `./vendor/bundle`
                           .                  __  ___
                   _, ,_  _| , . * ._   _    /  `  | 
                  (_| [ `(_] (_| | [ ) (_)   \__. _|_   v1.6.0

Host OS...                                                                 linux
Working directory...             /home/runner/work/TankController/TankController
Using configuration override from project...                     .arduino-ci.yml
Located arduino-cli binary...                           /home/runner/arduino-cli
Using arduino-cli version...                                              0.29.0
Found libraries directory...                      /home/runner/Arduino/libraries
Environment variable CUSTOM_INIT_SCRIPT...                                    ''
Environment variable USE_SUBDIR...                                            ''
Installing library under test...                                               ✓
Library installed at...            /home/runner/Arduino/libraries/TankController
bundler: failed to load command: arduino_ci.rb (/home/runner/work/TankController/TankController/vendor/bundle/ruby/3.2.0/bin/arduino_ci.rb)
Resolving 8 Arduino library dependencies configured in <library.properties>)... 
/home/runner/work/TankController/TankController/vendor/bundle/ruby/3.2.0/bundler/gems/arduino_ci-439d668458be/exe/arduino_ci.rb:167:in `block in install_arduino_library_dependencies_h': undefined method `library_of_name' for nil:NilClass (NoMethodError)

  (library_names.map { |n| @backend.library_of_name(n) } - installed).each do |l|
...Resolving 8 Arduino library dependencies configured in <library.properties>) 
                                   ^^^^^^^^^^^^^^^^
	from /home/runner/work/TankController/TankController/vendor/bundle/ruby/3.2.0/bundler/gems/arduino_ci-439d668458be/exe/arduino_ci.rb:167:in `map'
	from /home/runner/work/TankController/TankController/vendor/bundle/ruby/3.2.0/bundler/gems/arduino_ci-439d668458be/exe/arduino_ci.rb:167:in `install_arduino_library_dependencies_h'
	from /home/runner/work/TankController/TankController/vendor/bundle/ruby/3.2.0/bundler/gems/arduino_ci-439d668458be/exe/arduino_ci.rb:191:in `block in install_arduino_library_dependencies'
	from /home/runner/work/TankController/TankController/vendor/bundle/ruby/3.2.0/bundler/gems/arduino_ci-439d668458be/lib/arduino_ci/logger.rb:119:in `perform_action'
	from /home/runner/work/TankController/TankController/vendor/bundle/ruby/3.2.0/bundler/gems/arduino_ci-439d668458be/lib/arduino_ci/logger.rb:220:in `inform_multiline'
	from /home/runner/work/TankController/TankController/vendor/bundle/ruby/3.2.0/bundler/gems/arduino_ci-439d668458be/exe/arduino_ci.rb:190:in `install_arduino_library_dependencies'
	from /home/runner/work/TankController/TankController/vendor/bundle/ruby/3.2.0/bundler/gems/arduino_ci-439d668458be/exe/arduino_ci.rb:354:in `perform_bootstrap'
	from /home/runner/work/TankController/TankController/vendor/bundle/ruby/3.2.0/bundler/gems/arduino_ci-439d668458be/exe/arduino_ci.rb:605:in `<top (required)>'
	from /home/runner/work/TankController/TankController/vendor/bundle/ruby/3.2.0/bin/arduino_ci.rb:25:in `load'
	from /home/runner/work/TankController/TankController/vendor/bundle/ruby/3.2.0/bin/arduino_ci.rb:25:in `<top (required)>'
	from /opt/hostedtoolcache/Ruby/3.2.2/x64/lib/ruby/3.2.0/bundler/cli/exec.rb:58:in `load'
	from /opt/hostedtoolcache/Ruby/3.2.2/x64/lib/ruby/3.2.0/bundler/cli/exec.rb:58:in `kernel_load'
	from /opt/hostedtoolcache/Ruby/3.2.2/x64/lib/ruby/3.2.0/bundler/cli/exec.rb:23:in `run'
	from /opt/hostedtoolcache/Ruby/3.2.2/x64/lib/ruby/3.2.0/bundler/cli.rb:492:in `exec'
	from /opt/hostedtoolcache/Ruby/3.2.2/x64/lib/ruby/3.2.0/bundler/vendor/thor/lib/thor/command.rb:27:in `run'
	from /opt/hostedtoolcache/Ruby/3.2.2/x64/lib/ruby/3.2.0/bundler/vendor/thor/lib/thor/invocation.rb:127:in `invoke_command'
	from /opt/hostedtoolcache/Ruby/3.2.2/x64/lib/ruby/3.2.0/bundler/vendor/thor/lib/thor.rb:392:in `dispatch'
	from /opt/hostedtoolcache/Ruby/3.2.2/x64/lib/ruby/3.2.0/bundler/cli.rb:34:in `dispatch'
	from /opt/hostedtoolcache/Ruby/3.2.2/x64/lib/ruby/3.2.0/bundler/vendor/thor/lib/thor/base.rb:485:in `start'
	from /opt/hostedtoolcache/Ruby/3.2.2/x64/lib/ruby/3.2.0/bundler/cli.rb:28:in `start'
	from /opt/hostedtoolcache/Ruby/3.2.2/x64/lib/ruby/gems/3.2.0/gems/bundler-2.4.10/libexec/bundle:45:in `block in <top (required)>'
	from /opt/hostedtoolcache/Ruby/3.2.2/x64/lib/ruby/3.2.0/bundler/friendly_errors.rb:117:in `with_friendly_errors'
	from /opt/hostedtoolcache/Ruby/3.2.2/x64/lib/ruby/gems/3.2.0/gems/bundler-2.4.10/libexec/bundle:33:in `<top (required)>'
	from /opt/hostedtoolcache/Ruby/3.2.2/x64/bin/bundle:25:in `load'
	from /opt/hostedtoolcache/Ruby/3.2.2/x64/bin/bundle:25:in `<main>'
Error: Process completed with exit code 1.
```

--- a/.github/workflows/linux.yaml
+++ b/.github/workflows/linux.yaml
@@ -4,6 +4,21 @@ name: linux
 on: [push, pull_request]
 
 jobs:
+  "BusIO_with_dependencies":
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v3
+      - uses: ruby/setup-ruby@v1
+        with:
+          ruby-version: 2.6
+      - name: Check usage - Test BusIO from scratch
+        run: |
+          g++ -v
+          cd SampleProjects/BusIO
+          bundle install
+          bundle exec ensure_arduino_installation.rb
+          bundle exec arduino_ci.rb
+
   "rubocop":
     runs-on: ubuntu-latest
     steps:

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -15,6 +15,7 @@ and this project adheres to [Semantic Versioning](http://semver.org/).
 ### Removed
 
 ### Fixed
+- Fixes #352, in which the Arduino backend was being referenced in the CI runner before being explicitly assigned; this affected all tests of modules that used `library.properties` to specify runtime Arduino module dependencies.
 
 ### Security
 

--- a/SampleProjects/BusIO/.arduino-ci.yml
+++ b/SampleProjects/BusIO/.arduino-ci.yml
@@ -2,11 +2,10 @@ unittest:
   platforms:
     - mega2560
   libraries:
-    - "Adafruit BusIO"
-    # - "Adafruit_BusIO"    <= This works if you have the library pre-installed
+    - "Adafruit_BusIO"
 
 compile:
   platforms:
     - mega2560
   libraries:
-    - "Adafruit BusIO"
+    - "Adafruit_BusIO"

--- a/SampleProjects/BusIO/README.md
+++ b/SampleProjects/BusIO/README.md
@@ -1,4 +1,8 @@
 # BusIO
 
 This is an example of a library that depends on Adafruit BusIO.
-It is provided to help reproduce #192.
+It is provided to help reproduce #192 and #352.
+
+This example specifies a dependency in `library.properties`, which
+exercises the `arduino_ci.rb` CI runner in a way that the other
+SampleProjects currently do not.

--- a/SampleProjects/BusIO/library.properties
+++ b/SampleProjects/BusIO/library.properties
@@ -8,3 +8,4 @@ category=Other
 url=https://github.com/Arduino-CI/arduino_ci/SampleProjects/BusIO
 architectures=avr,esp8266
 includes=BusIO.h
+depends=Adafruit BusIO

--- a/exe/arduino_ci.rb
+++ b/exe/arduino_ci.rb
@@ -351,11 +351,6 @@ def perform_bootstrap
     end
   end
 
-  install_arduino_library_dependencies(
-    cpp_library.arduino_library_dependencies,
-    "<#{ArduinoCI::CppLibrary::LIBRARY_PROPERTIES_FILE}>"
-  )
-
   # return all objects needed by other steps
   {
     backend: backend,
@@ -604,6 +599,12 @@ end
 
 strap = perform_bootstrap
 @backend = strap[:backend]
+
+install_arduino_library_dependencies(
+  strap[:cpp_library].arduino_library_dependencies,
+  "<#{ArduinoCI::CppLibrary::LIBRARY_PROPERTIES_FILE}>"
+)
+
 perform_unit_tests(strap[:cpp_library], strap[:config])
 perform_example_compilation_tests(strap[:cpp_library], strap[:config])
 

